src/object/data.rs

```diff
@@ -0,0 +1,116 @@
+use std::cell::{Ref, RefCell};
+use std::iter::Iterator;
+use std::rc::Rc;
+
+/// View an object mutably.
+/// This trait drives the assignment into vectors and lists, primarily
+/// via the `Data` struct.
+pub trait ViewMut {
+    fn view_mut(&self) -> Self;
+}
+
+/// Internal Data representation for copy-on-write semantics.
+#[derive(Debug, PartialEq, Default)]
+pub struct Data<T: Clone>(Rc<RefCell<Rc<T>>>);
```

Comment:
Yeah, I agree! I think I would prefer `CowData` over `CowObj` to avoid any confusion with `Obj`.

Reply:
I think I do like having `Obj` in there. Maybe `ObjCowData`? At least for the purposes of this project, the intention behind this data structure is purely for managing copy-on-write of `Obj`'s, right? Or am I missing something?

I'm not sure that there's a clear winner here. The awkwardness of this convention is telling me that there might be a better way to model `Obj` altogether.

Maybe we need something like an `ObjRef` and an `ObjMut`?

Definitely not necessary for this PR, but generally I find that when I start having to dance around naming conventions, it's usually an early omen that some part of the data model is a bit wonky. It might take a few iterations for us to land on something that has the right feel.

Reply:
Ok, let's just go with `CowObj`! 


Regarding the remodelling:
One thought I had was whether it would make sense do define `Obj` as a trait instead of a enum, so if we think about changing the data model I would try to explore that option. I feel the it could make code more readable by bundling code for each object type separately instead of having the code for all object type in the methods of `Obj`. Coming up with a proper trait definition for an object would also mean that it would be possible for others to extend the language with new objects by implementing that trait.

Reply:
The trait style would be interesting. My (admittedly naive) understanding is that traits can come at a performance cost because each function needs to go through a vtable to do dynamic dispatch. Going through a vtable also means that the compiler won't do any extra optimizing at the function boundary.

I don't really have a strong intuition for exactly how steep that cost will be, though. In a language like R, maybe that performance cost is negligible.

Reply:
I have read that in most cases the cost for dynamic dispatch is negligible (unless it is e.g. a hot loop). But I also don't really have an intuition whether a central data structure for the language counts as such an exception. 

If I can dream a little, I think the way to get performance eventually would be to adapt a simple type system that then allows to compile what you here call "static" functions: https://github.com/dgkf/R/issues/62.

Reply:
Yeah - that's a good point. I think it's worth a little exploration, but if it looks like it's going to be a huge lift, there are a few higher priority things I'd like to suss out first. Most notably, the next big priority feature would be async functions and I have a feeling that might be easier to start with a simple enum so I can avoid having to worry about send/sync implementations.

Reply:
Ok, I will but this off for the time being!